core/src/main/java/com/oath/oak/InternalOakMap.java

```diff
@@ -513,42 +497,49 @@ void putIfAbsentComputeIfPresent(K key, V value, Consumer<OakWBuffer> computer)
         if (ei == -1) {
             ei = c.allocateEntryAndKey(key);
             if (ei == -1) {
-                rebalanceCompute(c, key, value, computer);
-                return;
+                rebalance(c);
+                return putIfAbsentComputeIfPresent(key, value, computer);
             }
             int prevEi = c.linkEntry(ei, true, key);
             if (prevEi != ei) {
-                Handle handle = c.getHandle(prevEi);
-                if (handle == null) {
-                    ei = prevEi;
-                    prevHi = c.getHandleIndex(prevEi);
-                } else {
-                    if (handle.compute(computer)) {
+                prevHi = c.getHandleIndex(prevEi);
+                if (prevHi != -1) {
+                    if (c.getHandle(prevEi).compute(computer)) {
                         // compute was successful and handle wasn't found deleted; in case
                         // this handle was already found as deleted, continue to construct another handle
-                        return;
+                        return false;
                     }
+                } else {
+                    ei = prevEi;
                 }
             }
         }
 
         int hi = c.allocateHandle();
         if (hi == -1) {
-            rebalanceCompute(c, key, value, computer);
-            return;
+            rebalance(c);
+            return putIfAbsentComputeIfPresent(key, value, computer);
         }
 
         c.writeValue(hi, value); // write value in place
 
         Chunk.OpData opData = new Chunk.OpData(Operation.COMPUTE, ei, hi, prevHi, computer);
 
         // publish put
-        if (!c.publish(opData)) {
-            rebalanceCompute(c, key, value, computer);
-            return;
+        if (!c.publish()) {
+            c.freeHandle(hi);
```

Comment:
what happens if you do not free the handle - this is a memory leak?

Reply:
yes